File: rebus-rebuscrus/src/java/it/csi/rebus/rebuscrus/business/service/ContribuzioneService.java
/*
* SPDX-FileCopyrightText: (C) Copyright 2022 Regione Piemonte
* SPDX-License-Identifier: EUPL-1.2
*/
package it.csi.rebus.rebuscrus.business.service;

import java.util.HashMap;
import java.util.List;

import it.csi.rebus.rebuscrus.integration.dto.VExportRicercaContribuzioneDTO;
import it.csi.rebus.rebuscrus.vo.AssegnazioneRisorseVO;
import it.csi.rebus.rebuscrus.vo.AttoAssegnazioneRisorseVO;
import it.csi.rebus.rebuscrus.vo.ContribuzioneCompletaVO;
import it.csi.rebus.rebuscrus.vo.ContribuzioneVO;
import it.csi.rebus.rebuscrus.vo.CostoFornituraVO;
import it.csi.rebus.rebuscrus.vo.DatoBonificoVO;
import it.csi.rebus.rebuscrus.vo.DatoFatturaVO;
import it.csi.rebus.rebuscrus.vo.DatoMessaServizioVO;
import it.csi.rebus.rebuscrus.vo.DatoSpesaVO;
import it.csi.rebus.rebuscrus.vo.DocContribuzioneVO;
import it.csi.rebus.rebuscrus.vo.DocDatoMessaServizioVO;
import it.csi.rebus.rebuscrus.vo.FiltroContribuzioneVO;
import it.csi.rebus.rebuscrus.vo.FonteFinanziamentoVO;
import it.csi.rebus.rebuscrus.vo.InserisciRichiestaVO;
import it.csi.rebus.rebuscrus.vo.OrdineAcquistoVO;
import it.csi.rebus.rebuscrus.vo.ResponseStringVO;
import it.csi.rebus.rebuscrus.vo.TipoDocumentoQuietanzaVO;
import it.csi.rebus.rebuscrus.vo.TipoSostituzioneVO;
import it.csi.rebus.rebuscrus.vo.VoceCostoFornituraVO;
import it.csi.rebus.rebuscrus.vo.VoceDiCostoContribuzioneVO;

public interface ContribuzioneService {

	List<VoceDiCostoContribuzioneVO> getAllVoceCostoContribuzione();

	List<AttoAssegnazioneRisorseVO> getAllAttoAssegnazioneRisorse();

	List<TipoDocumentoQuietanzaVO> getAllTipoDocumentoQuietanza();

	List<TipoSostituzioneVO> getAllTipoSostituzione();

	List<FonteFinanziamentoVO> getAllFonteFinanziamentoByIdAttoAssegnazione(Long idAttoAssegnazione);

	List<FonteFinanziamentoVO> getAllFonteFinanziamento();

	// Sezione: Assegnazione delle risorse - Contributi erogati
	Long insertAssegnazioneRisorse(AssegnazioneRisorseVO assegnazioneRisorse);

	// Sezione: Ordine di acquisto (GARA)
	Long insertOrdineAcquisto(OrdineAcquistoVO ordineAcquisto);

	// Sezione: Voci di costo delle forniture
	HashMap<Long, Long> insertVoceCostoFornitura(List<VoceCostoFornituraVO> voceCostoFornitura, Long idCostoFornitura);

	Long insertCostoFornitura(CostoFornituraVO costoFornitura, Long idDocCostoFornitura);

	// Sezione: Contributi erogabili e documentazione di spesa e di pagamento

	Long insertDatoSpesa(DatoSpesaVO datoSpesa);

	Long insertDatoFattura(DatoFatturaVO datoFattura, Long idDoc);

	void insertOggettoFattura(Long idDatoFattura, Long idVoceCostoFornitura);

	// Sezione: Messa in servizio, rispetto dei vincoli e all assolvimento delle
	// azioni di pubblicita

	Long insertDatoMessaServizio(DatoMessaServizioVO datoMessaServizio);

	void insertDocDatoMessaServizio(Long idDatoMessaServizio, Long idDocumento);

	// Sezione: Stato della rendicontazione
	Long insertContribuzione(ContribuzioneCompletaVO contribuzioneCompleta, InserisciRichiestaVO richiestaRequest);

	Long insertDocContribuzione(DocContribuzioneVO datiDocumento);

	DocContribuzioneVO getDocumentoById(Long id);

	List<VExportRicercaContribuzioneDTO> filtraElencoContribuzione(FiltroContribuzioneVO filtro);

	List<VExportRicercaContribuzioneDTO> elencoContribuzione();

	// GET CONTRIBUZIONE
	ContribuzioneVO getContribuzioneByIdProcedimento(Long idProcedimento);

	ContribuzioneVO getContribuzioneById(Long idContribuzione);

	// Sezione: Assegnazione delle risorse - Contributi erogati
	AssegnazioneRisorseVO getAssegnazioneRisorseById(Long idAassegnazioneRisorse);

	// Sezione: Ordine di acquisto (GARA)
	OrdineAcquistoVO getOrdineAcquistoById(Long idOrdineAcquisto);

	// Sezione: Voci di costo delle forniture
	CostoFornituraVO getCostoFornituraById(Long idCostoFornitura);

	List<VoceCostoFornituraVO> getVoceCostoFornituraByIdCostoFornitura(Long idCostoFornitura);

	// RESTITUISCE LE INFO DEL DOCUMENTO SENZA IL FILE
	DocContribuzioneVO getInfoDocContribuzione(Long idDocumento);

	// RESTITUISCE IL DOUMENTO PER IL DOWNLOAD (CHIAMATA SUL BOTTONE DI DOWNLOAD)
	byte[] getDocContribuzione(Long idDocumento);

	// Sezione: Contributi erogabili e documentazione di spesa e di pagamento

	DatoSpesaVO getDatoSpesaById(Long idDatoSpesa);

	List<DatoFatturaVO> getDatoFatturaByIdDatoSpesa(Long idDatoSpesa);

	List<Long> getOggettoFatturaByIdDatoFattura(Long idDatoFattura);

	// Sezione: Messa in servizio, rispetto dei vincoli e all assolvimento delle
	// azioni di pubblicita

	DatoMessaServizioVO getDatoMessaServizioById(Long idDatoMessaServizio);

	List<DocDatoMessaServizioVO> getDocDatoMessaServizioByIdDatoMessaServizio(Long idDatoMessaServizio);

	ContribuzioneCompletaVO getContribuzioneCompletaByIdProcedimento(Long idProcedimento);

	ContribuzioneCompletaVO getContribuzioneCompletaById(Long idContribuzione);

	// Sezione: Assegnazione delle risorse - Contributi erogati
	Long updateAssegnazioneRisorse(AssegnazioneRisorseVO assegnazioneRisorse);

	// Sezione: Ordine di acquisto (GARA)
	Long updateOrdineAcquisto(OrdineAcquistoVO ordineAcquisto);

	Long updateCostoFornitura(CostoFornituraVO costoFornitura);

	Long updateContribuzione(ContribuzioneCompletaVO contribuzione, List<Long> listaVociCostoDaEliminare,
			List<Long> listaDatiFatturaDaEliminare, List<Long> listaDatiBonificoDaEliminare);

	// Sezione: Contributi erogabili e documentazione di spesa e di pagamento

	Long updateDatoSpesa(DatoSpesaVO datoSpesa);

	// Sezione: Messa in servizio, rispetto dei vincoli e all assolvimento delle
	// azioni di pubblicita

	Long updateDatoMessaServizio(DatoMessaServizioVO datoMessaServizio);
	
	Long updateDocContribuzione(DocContribuzioneVO datiDocumento);

	void updateDocDatoMessaServizio(Long oldDocumentId, DocContribuzioneVO newDocument, Long IdDatoMessaServizio);

	ResponseStringVO getTelaioByIdProcedimento(Long IdProcediemnto);
	
	boolean checkFinalStateIter(Long idStato);
	
	void insertDatoBonificoAndFatturaBonifico(List<DatoBonificoVO> listBonifici, Long idDatoFattura);
	
	List<String> getTelaiVeicoloDaSostituire();
}

File: rebus-rebuscruswcl/src/app/services/contribuzione.service.ts
/***************************************************
Copyright Regione Piemonte - 2022
 *SPDX-License-Identifier: EUPL-1.2-or-later
***************************************************/
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs/Observable";
import { AttoAssegnazioneRisorseVo } from "../vo/contribuzione/atto-assegnazione-risorse-vo";
import { ContribuzioneCompletaVO } from "../vo/contribuzione/contribuzione-completa-vo";
import { FonteFinanziamentoVO } from "../vo/contribuzione/fonte-finanziamento-vo";
import { TipoDocumentoQuietanzaVO } from "../vo/contribuzione/tipo-documento-quietanza-vo";
import { TipoSostituzioneVO } from "../vo/contribuzione/tipo-sostituzione-vo";
import { VoceDiCostoContribuzioneVO } from "../vo/contribuzione/voce-costo-contribuzione-vo";
import { FiltroContribuzioneVO } from "../vo/filtro-contribuzione-vo";
import { InserisciRichiestaVO } from "../vo/inserisci-richiesta.vo";

@Injectable()
export class ContribuzioneService {

    private urlContribuzione: string = "/parcobustplweb/restfacade/contribuzione";

    primoTelaio: string;
    idProcedimento: number;
    idVariazAutobus: number;

    constructor(private http: HttpClient) {

    }

    //************************* UTILITY *************************//

    resetFields() {
        this.primoTelaio = undefined;
        this.idProcedimento = undefined;
        this.idVariazAutobus = undefined;
    }

    checkFinalStateIter(idStato: number): Observable<any> {
        let params = new HttpParams().set('idStato', JSON.stringify(idStato))
        return this.http.get<any>(this.urlContribuzione + "/checkFinalStateIter", { params: params });
    }

    //************************* GET *************************//

    //Get DropDown Voci costo contribuazione
    getAllVoceCostoContribuzione(): Observable<VoceDiCostoContribuzioneVO[]> {
        return this.http.get<VoceDiCostoContribuzioneVO[]>(this.urlContribuzione + "/getAllVoceCostoContribuzione");
    }

    //Get DropDown Atto assegnazione
    getAllAttoAssegnazioneRisorse(): Observable<AttoAssegnazioneRisorseVo[]> {
        return this.http.get<AttoAssegnazioneRisorseVo[]>(this.urlContribuzione + "/getAllAttoAssegnazioneRisorse");
    }

    //Get DropDown tipo documento quietanza
    getAllTipoDocumentoQuietanza(): Observable<TipoDocumentoQuietanzaVO[]> {
        return this.http.get<TipoDocumentoQuietanzaVO[]>(this.urlContribuzione + "/getAllTipoDocumentoQuietanza");
    }

    //Get DropDown tipo sostituzione
    getAllTipoSostituzione(): Observable<TipoSostituzioneVO[]> {
        return this.http.get<TipoSostituzioneVO[]>(this.urlContribuzione + "/getAllTipoSostituzione");
    }

    //Get DropDown tipo sostituzione
    getAllFonteFinanziamentoByIdAttoAssegnazione(idAttoAssegnazione: number): Observable<FonteFinanziamentoVO[]> {
        let params = new HttpParams().set('idAttoAssegnazione', JSON.stringify(idAttoAssegnazione))
        return this.http.get<FonteFinanziamentoVO[]>(this.urlContribuzione + "/getAllFonteFinanziamentoByIdAttoAssegnazione", { params: params });
    }

    //Get DropDown fonte di finanziamento
    getAllFonteFinanziamento(): Observable<FonteFinanziamentoVO[]> {
        return this.http.get<FonteFinanziamentoVO[]>(this.urlContribuzione + "/getAllFonteFinanziamento");
    }
    getContribuzioneCompletaByIdProcedimento(idProcedimento: number): Observable<ContribuzioneCompletaVO> {
        let params = new HttpParams().set('idProcedimento', JSON.stringify(idProcedimento))
        return this.http.get<ContribuzioneCompletaVO>(this.urlContribuzione + "/getContribuzioneCompletaByIdProcedimento", { params: params });
    }

    getContribuzioneCompletaById(idContribuzione: number): Observable<ContribuzioneCompletaVO> {
        let params = new HttpParams().set('idContribuzione', JSON.stringify(idContribuzione))
        return this.http.get<ContribuzioneCompletaVO>(this.urlContribuzione + "/getContribuzioneCompletaById", { params: params });
    }

    getDocContribuzione(idDoc: number) {
        let params = new HttpParams().set('idDoc', JSON.stringify(idDoc))
        return this.http.get(this.urlContribuzione + "/getDocContribuzione", { params: params, responseType: 'blob' }).map(
            (res) => {
                return new Blob([res], { type: 'application/pdf' })
            });
    }

    getTelaioByIdProcedimento(idProcedimento: number): Observable<any> {
        let params = new HttpParams().set('idProcedimento', JSON.stringify(idProcedimento))
        return this.http.get<any>(this.urlContribuzione + "/getTelaioByIdProcedimento", { params: params });
    }

    getTelaiVeicoloDaSostituire(): Observable<any> {
        return this.http.get<any>(this.urlContribuzione + "/getTelaiVeicoloDaSostituire");
    }

    //************************* INSERT *************************//

    insertContribuzione(contribuzione: ContribuzioneCompletaVO, richiesta: InserisciRichiestaVO): Observable<number> {
        var form = new FormData();
        form.append('contribuzione', JSON.stringify(contribuzione))
        form.append('richiesta', JSON.stringify(richiesta));
        return this.http.post<number>(this.urlContribuzione + "/inserisciContribuzione", form);
    }

    //************************* UPDATE *************************//

    updateContribuzione(contribuzione: ContribuzioneCompletaVO, listOfDeletedIdVociCosto: number[], listOfDeletedIdDatiFattura: number[], listOfDeletedIdDatiBonifico: number[]) {
        var form = new FormData();
        form.append('contribuzione', JSON.stringify(contribuzione))
        form.append('listaVociCostoDaEliminare', JSON.stringify(listOfDeletedIdVociCosto));
        form.append('listaDatiFatturaDaEliminare', JSON.stringify(listOfDeletedIdDatiFattura));
        form.append('listaDatiBonificoDaEliminare', JSON.stringify(listOfDeletedIdDatiBonifico));
        return this.http.post<number>(this.urlContribuzione + "/updateContribuzione", form);
    }

    //************************* UPDATE *************************//

    elencoContribuzione() {
        let params = new HttpParams();
        return this.http.get<Array<FiltroContribuzioneVO>>(this.urlContribuzione + "/elencoContribuzione", { params: params });
    }



    filtraElencoContribuzione(filtro: FiltroContribuzioneVO) {
        const body = filtro;
        return this.http.post<Array<FiltroContribuzioneVO>>(this.urlContribuzione + "/filtraElencoContribuzione", body);
    }

}